app/src/main/java/com/randomappsinc/studentpicker/common/SpeechToTextManager.java

```diff
@@ -0,0 +1,127 @@
+package com.randomappsinc.studentpicker.common;
+
+import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+import android.speech.RecognitionListener;
+import android.speech.RecognizerIntent;
+import android.speech.SpeechRecognizer;
+import android.view.View;
+import android.widget.TextView;
+import com.afollestad.materialdialogs.MaterialDialog;
+import com.joanzapata.iconify.widget.IconTextView;
+import com.randomappsinc.studentpicker.R;
+import java.util.List;
+import butterknife.BindView;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+
+public class SpeechToTextManager implements RecognitionListener {
+    public interface Listener {
+        void onSpokenText(String spokenText);
+    }
+
+    @BindView(R.id.voice_icon) IconTextView voiceIcon;
+    @BindView(R.id.message) TextView message;
+    @BindView(R.id.try_again) View tryAgain;
+
+    private SpeechRecognizer speechRecognizer;
+    private Intent speechRecognizerIntent;
+    private Listener listener;
+    private MaterialDialog dialog;
+
+    public SpeechToTextManager(Context context, Listener listener) {
+        this.listener = listener;
+        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(context);
+        speechRecognizerIntent = new Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH);
+        speechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE_MODEL, RecognizerIntent.LANGUAGE_MODEL_WEB_SEARCH);
+        speechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_LANGUAGE, "en");
+        speechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true);
+        speechRecognizerIntent.putExtra(RecognizerIntent.EXTRA_CALLING_PACKAGE, context.getPackageName());
+        speechRecognizer.setRecognitionListener(this);
+
+        dialog = new MaterialDialog.Builder(context)
+                .customView(R.layout.voice_recognizble_layout, false)
+                .build();
+        ButterKnife.bind(this, dialog);
+    }
+
+    public void showSpeechDialog() {
```

Comment:
Speech to text having UI is another thing we should hide from the caller. The caller should just know if I use this module and start the flow, I will eventually get back a string (assuming all goes well). It should be complete magic; the caller shouldn't even know there's a dialog somewhere in the mix. This API is also inaccurate, because it doesn't just show a dialog, it starts listening to the user, making the 2nd line feel like more of a side effect. Let's change this API to the suggestion below. Also, we still need to take in a customizable message resource, similar to what our implementation did before.
```suggestion
    public void startSpeechToTextFlow(@StringRes int messageId) {
```

Reply:
The message we send to startSpeechToTextFlow is the one which is gonna be shown on the dialog?
If yes then I think it's better to make changeUIStateToListening, changeUIStateToRetry to only one function changeUIStates, then send the variables for it to make it perform as we wish!

Reply:
startSpeechToTextFlow API should have 3 purposes:
1- Showing SpeechToTextDialog
2- Start listening
3- Changing UI states to listening